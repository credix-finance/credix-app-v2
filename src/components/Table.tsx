/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useEffect, useState } from "react";
import { Table as AntdTable } from "antd";
import { TableProps as AntdTableProps } from "antd/lib/table";
import { PageItem } from "@components/PageItem";
import { ColumnType } from "antd/lib/table";
import { TableHeaderCell } from "./TableHeaderCell";
import { IconName } from "@components/Icon";

type ColumnsProps = ColumnType<any> & {
	/**
	 * Optional icon that will be displayed on the left of the column title
	 */
	icon?: IconName;
	titleClassName?: string;
};

interface TableProps {
	/**
	 * The data which will fill the table
	 */
	dataSource?: AntdTableProps<any>["dataSource"];
	/**
	 * Table columns
	 */
	columns?: ColumnsProps[];
	/**
	 * onRow provides a way to hook into click events originating from the table row.
	 * This prop also enables table hover styles
	 */
	onRow?: AntdTableProps<any>["onRow"];
}

export const Table = ({ columns, ...props }: TableProps) => {
	const [parsedColumns, setParsedColumns] = useState([]);

	useEffect(() => {
		setParsedColumns(
			columns.map((column) => {
				if (typeof column.title === "function") {
					return column;
				}

				if (column.icon) {
					const { icon, title, titleClassName } = column;
					return Object.assign({}, column, {
						title: () => (
							<TableHeaderCell label={title as string} icon={icon} className={titleClassName} />
						),
					});
				}

				return column;
			})
		);
	}, [columns]);

	return (
		<AntdTable
			className={props.onRow ? "table-hoverable" : ""}
			pagination={{
				hideOnSinglePage: true,
				pageSize: 5,
				itemRender: (page, type, originalElement) => (
					<PageItem page={page} type={type} originalElement={originalElement} />
				),
			}}
			columns={parsedColumns}
			{...props}
		/>
	);
};
