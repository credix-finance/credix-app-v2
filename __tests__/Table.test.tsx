import React from "react";
import renderer from "react-test-renderer";
import { Table } from "@components/Table";

const defaultData = {
	dataSource: [
		{
			key: 0,
			name: "a55 Debenture Series-1",
			amount: "0.22 USDC",
			date: "02.11.21",
		},
		{
			key: 1,
			name: "a55 Debenture Series-1",
			amount: "0.22 USDC",
			date: "02.11.21",
		},
		{
			key: 2,
			name: "a55 Debenture Series-1",
			amount: "0.22 USDC",
			date: "02.11.21",
		},
		{
			key: 3,
			name: "a55 Debenture Series-1",
			amount: "0.22 USDC",
			date: "02.11.21",
		},
		{
			key: 4,
			name: "a55 Debenture Series-1",
			amount: "0.22 USDC",
			date: "02.11.21",
		},
	],
	columns: [
		{
			title: "Name",
			dataIndex: "name",
			key: "name",
		},
		{
			title: "Amount",
			dataIndex: "amount",
			key: "amount",
			render: (text) => <span className="font-medium text-lg">{text}</span>,
		},
		{
			title: "Date",
			dataIndex: "date",
			key: "date",
			render: (text) => <span className="font-medium text-lg">{text}</span>,
		},
	],
};

test("Default", () => {
	const component = renderer.create(<Table {...defaultData} />);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test("Table headers with icons", () => {
	const columns = [
		{
			title: "Name",
			icon: "stacked-column-down",
			dataIndex: "name",
			key: "name",
		},
		{
			title: "Amount",
			icon: "stacked-column-down",
			dataIndex: "amount",
			key: "amount",
			render: (text) => <span className="font-medium text-lg">{text}</span>,
		},
		{
			title: "Date",
			icon: "calendar",
			dataIndex: "date",
			key: "date",
			render: (text) => <span className="font-medium text-lg">{text}</span>,
		},
	];

	const data = Object.assign({}, defaultData, { columns });

	const component = renderer.create(<Table {...data} />);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
