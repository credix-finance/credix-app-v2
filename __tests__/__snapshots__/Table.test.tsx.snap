// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default 1`] = `
<div
  className="ant-table-wrapper"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table"
      >
        <div
          className="ant-table-container"
        >
          <div
            className="ant-table-content"
            onScroll={[Function]}
            style={Object {}}
          >
            <table
              style={
                Object {
                  "tableLayout": "auto",
                }
              }
            >
              <colgroup />
              <thead
                className="ant-table-thead"
              >
                <tr>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={0}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={1}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={2}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={3}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={4}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  />
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table headers with icons 1`] = `
<div
  className="ant-table-wrapper"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table"
      >
        <div
          className="ant-table-container"
        >
          <div
            className="ant-table-content"
            onScroll={[Function]}
            style={Object {}}
          >
            <table
              style={
                Object {
                  "tableLayout": "auto",
                }
              }
            >
              <colgroup />
              <thead
                className="ant-table-thead"
              >
                <tr>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="space-x-1 flex items-center "
                    >
                      <svg
                        className="w-4 h-4 fill-current"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clipRule="evenodd"
                          d="M12 7.333a.667.667 0 1 0 0 1.334h1.333a.667.667 0 1 0 0-1.334H12Zm-.667 3.334c0-.368.299-.667.667-.667h1.333a.667.667 0 0 1 0 1.333H12a.667.667 0 0 1-.667-.666Zm0 2.666c0-.368.299-.666.667-.666h1.333a.667.667 0 0 1 0 1.333H12a.667.667 0 0 1-.667-.667ZM7.334 4.667a.667.667 0 1 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334ZM6.667 8c0-.368.298-.667.667-.667h1.333a.667.667 0 0 1 0 1.334H7.334A.667.667 0 0 1 6.667 8Zm.667 2a.667.667 0 1 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334Zm0 2.667a.667.667 0 0 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334ZM2.667 2a.667.667 0 0 0 0 1.333H4A.667.667 0 1 0 4 2H2.667ZM2 5.333c0-.368.298-.666.667-.666H4A.667.667 0 1 1 4 6H2.667A.667.667 0 0 1 2 5.333Zm.667 2a.667.667 0 0 0 0 1.334H4a.667.667 0 0 0 0-1.334H2.667ZM2 10.667c0-.368.298-.667.667-.667H4a.667.667 0 0 1 0 1.333H2.667A.667.667 0 0 1 2 10.667Zm0 2.666c0-.368.298-.666.667-.666H4A.667.667 0 1 1 4 14H2.667A.667.667 0 0 1 2 13.333Z"
                          fillRule="evenodd"
                        />
                      </svg>
                      <span>
                        Name
                      </span>
                    </div>
                  </th>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="space-x-1 flex items-center "
                    >
                      <svg
                        className="w-4 h-4 fill-current"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clipRule="evenodd"
                          d="M12 7.333a.667.667 0 1 0 0 1.334h1.333a.667.667 0 1 0 0-1.334H12Zm-.667 3.334c0-.368.299-.667.667-.667h1.333a.667.667 0 0 1 0 1.333H12a.667.667 0 0 1-.667-.666Zm0 2.666c0-.368.299-.666.667-.666h1.333a.667.667 0 0 1 0 1.333H12a.667.667 0 0 1-.667-.667ZM7.334 4.667a.667.667 0 1 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334ZM6.667 8c0-.368.298-.667.667-.667h1.333a.667.667 0 0 1 0 1.334H7.334A.667.667 0 0 1 6.667 8Zm.667 2a.667.667 0 1 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334Zm0 2.667a.667.667 0 0 0 0 1.333h1.333a.667.667 0 0 0 0-1.333H7.334ZM2.667 2a.667.667 0 0 0 0 1.333H4A.667.667 0 1 0 4 2H2.667ZM2 5.333c0-.368.298-.666.667-.666H4A.667.667 0 1 1 4 6H2.667A.667.667 0 0 1 2 5.333Zm.667 2a.667.667 0 0 0 0 1.334H4a.667.667 0 0 0 0-1.334H2.667ZM2 10.667c0-.368.298-.667.667-.667H4a.667.667 0 0 1 0 1.333H2.667A.667.667 0 0 1 2 10.667Zm0 2.666c0-.368.298-.666.667-.666H4A.667.667 0 1 1 4 14H2.667A.667.667 0 0 1 2 13.333Z"
                          fillRule="evenodd"
                        />
                      </svg>
                      <span>
                        Amount
                      </span>
                    </div>
                  </th>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="space-x-1 flex items-center "
                    >
                      <svg
                        className="w-4 h-4 fill-current"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clipRule="evenodd"
                          d="M8.759 5h6.482c.805 0 1.47 0 2.01.044.563.046 1.08.145 1.565.392a4 4 0 0 1 1.748 1.748c.247.485.346 1.002.392 1.564.044.541.044 1.206.044 2.01v4.483c0 .805 0 1.47-.044 2.01-.046.563-.145 1.08-.392 1.565a4 4 0 0 1-1.748 1.748c-.485.247-1.002.346-1.564.392-.541.044-1.206.044-2.01.044H8.758c-.805 0-1.47 0-2.01-.044-.563-.046-1.08-.145-1.565-.392a4 4 0 0 1-1.748-1.748c-.247-.485-.346-1.002-.392-1.564C3 16.71 3 16.046 3 15.242v-4.483c0-.805 0-1.47.044-2.01.046-.563.145-1.08.392-1.565a4 4 0 0 1 1.748-1.748c.485-.247 1.002-.346 1.564-.392C7.29 5 7.954 5 8.758 5ZM6.91 7.038c-.438.035-.663.1-.819.18a2 2 0 0 0-.874.874c-.08.156-.145.38-.18.819C5 9.361 5 9.943 5 10.8v4.4c0 .857 0 1.439.038 1.889.035.438.1.663.18.819a2 2 0 0 0 .874.874c.156.08.38.145.819.18C7.361 19 7.943 19 8.8 19h6.4c.857 0 1.439 0 1.889-.038.438-.035.663-.1.819-.18a2 2 0 0 0 .874-.874c.08-.156.145-.38.18-.819.037-.45.038-1.032.038-1.889v-4.4c0-.857 0-1.439-.038-1.889-.035-.438-.1-.663-.18-.819a2 2 0 0 0-.874-.874c-.156-.08-.38-.145-.819-.18C16.639 7 16.057 7 15.2 7H8.8c-.857 0-1.439 0-1.889.038Z"
                          fillRule="evenodd"
                        />
                        <path
                          d="M9 4a1 1 0 0 0-2 0v2a1 1 0 0 0 2 0V4ZM17 4a1 1 0 1 0-2 0v2a1 1 0 1 0 2 0V4Z"
                        />
                        <path
                          clipRule="evenodd"
                          d="M8 12a1 1 0 1 0 0-2 1 1 0 0 0 0 2Zm5-1a1 1 0 1 1-2 0 1 1 0 0 1 2 0Zm4 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0Zm-4 4a1 1 0 1 1-2 0 1 1 0 0 1 2 0Zm-5 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
                          fillRule="evenodd"
                        />
                      </svg>
                      <span>
                        Date
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={0}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    a55 Debenture Series-1
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      0.22 USDC
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      02.11.21
                    </span>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={1}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    a55 Debenture Series-1
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      0.22 USDC
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      02.11.21
                    </span>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={2}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    a55 Debenture Series-1
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      0.22 USDC
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      02.11.21
                    </span>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={3}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    a55 Debenture Series-1
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      0.22 USDC
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      02.11.21
                    </span>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={4}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    a55 Debenture Series-1
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      0.22 USDC
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      className="font-medium text-lg"
                    >
                      02.11.21
                    </span>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
